File: ts/src/handlers/http/allow_methods.test.ts
import { FromCommandLine } from "../../app.js";
import { MockApiHandler, MockHttpRequest, MockResponseWriter } from "../../testutils.js";
import { HttpMethods, HttpStatusCode } from "../../types/index.js";
import { AllowMethods } from "./allow_methods.js";

describe("TestAllowMethod", () => {
  it("AllowMethodEmpty", async () => {
    const hdl = new MockApiHandler<never, never>(FromCommandLine([]), MockHttpRequest());
    const fn = AllowMethods();
    expect(await fn(hdl)).toBeTruthy();
  });

  it("AllowMethodNotHit", async () => {
    const fn = AllowMethods("TEST", "REST");
    const hdl = new MockApiHandler<never, never>(
      FromCommandLine([]),
      MockHttpRequest({
        Method: "BRETT" as HttpMethods,
      }),
    );
    expect(await fn(hdl)).toBeFalsy();
    expect((hdl.Response() as MockResponseWriter).StatusCode).toBe(HttpStatusCode.METHOD_NOT_ALLOWED);
  });
  it("AllowMethodMatch", async () => {
    const methods = ["TEST", "REST"];
    const fn = AllowMethods(...methods);
    for (const method of methods) {
      const hdl = new MockApiHandler<never, never>(
        FromCommandLine([]),
        MockHttpRequest({
          Method: method as HttpMethods,
        }),
      );
      expect(await fn(hdl)).toBeTruthy();
      expect((hdl.Response() as MockResponseWriter).StatusCode).toBe(0);
    }
  });
});

File: go/handlers/http/allow_methods_test.go
package handlers_http

import (
	"net/http"
	"testing"

	"github.com/mabels/runified/testutils"
)

func TestAllowMethodEmpty(t *testing.T) {
	fn := AllowMethods()
	apih := &testutils.MockApiHandler[struct{}, struct{}]{
		Res: testutils.NewMockResponseWriter(),
	}
	fn(apih)
	if apih.Res.StatusCode == 405 {
		t.Error("Status code not set to 405", apih.Res.StatusCode)
	}
}

func TestAllowMethodNotHit(t *testing.T) {
	fn := AllowMethods("TEST", "REST")
	apih := &testutils.MockApiHandler[struct{}, struct{}]{
		Res: testutils.NewMockResponseWriter(),
		Req: &http.Request{
			Method: "BRETT",
		},
	}
	fn(apih)
	if apih.Res.StatusCode != 405 {
		t.Error("Status code not set to 405", apih.Res.StatusCode)
	}
}

func TestAllowMethodMatch(t *testing.T) {
	methods := []string{"TEST", "REST"}
	fn := AllowMethods(methods...)
	for _, method := range methods {
		apih := &testutils.MockApiHandler[struct{}, struct{}]{
			Res: testutils.NewMockResponseWriter(),
			Req: &http.Request{
				Method: method,
			},
		}
		fn(apih)
		if apih.Res.StatusCode == 405 {
			t.Error("Status code not set to 405", apih.Res.StatusCode)
		}
	}
}
